### ISSUE TITLE
A Question Concerning the Replacing of Similarity Matrix

### ISSUE THREAD

**Issue by cuilongyin** [NicolasHug/Surprise#301]
#### Description

Hi dear Nicolas:

About 8 days ago I asked a question about replacing the similarity matrix (#300 ). 
As you said, I hard set the matrix to a computed matrix instead of calling the function "self.compute_similarity"

However, I realize that the similarity matrix is playing a very trivial role in RMSE or MAE. I was thinking that because the variable self.sim is directly affecting the k_neighbors and thus affecting the values of sum_ratings / sum_sim, it must be very important in predicting the ratings in the testSet. But the result says otherwise, I wonder if I am missing something or there are other modifications I need to make to let the predictions entirely depend on the pre-computed matrix.

So, I tried filling the matrix with all ones, zeros and even random large integers and the result is different but not bad at all. The RMSE is around 1.0. But for a random recommender, I remember it is usually around 1.4.

To avoid making some amatrue coding mistakes, I tried copy-pasting the entire code from the file knns.py and changing only the fit function. Then, instead of importing KNNBasic from surprise I imported KNNBasic from that local file. But that hasn't change anything ...

I understand this question might be a bit off-topic, but if would be great if you could share some insights...

Thank you so much

#### Steps/Code to Reproduce

from MovieLens import MovieLens # A class I used to load up the dataset
from newKNN import KNNBasic
from surprise.model_selection import train_test_split
from surprise import accuracy

ml = MovieLens()
data = ml.loadMovieLensLatestSmall()
trainSet, testSet = train_test_split(data, test_size=.2, random_state=1)

sim_options = {'name': 'cosine',
               'user_based': True
               }

model = KNNBasic(sim_options=sim_options)
model.fit(trainSet)
predictions = model.test(testSet)

MAE = accuracy.mae(predictions, verbose=True)
RMSE = accuracy.rmse(predictions, verbose=True)

#And in the KNNBasic class, I only changed one line. 
   def fit(self, trainset):

        SymmetricAlgo.fit(self, trainset)      
        self.sim = np.ones(shape=(671,671), dtype = float, order = 'C')
        return self

#### Expected Results
    Some large RMSE or MAE
#### Actual Results
MAE:  0.7719
RMSE: 0.9925
#### Versions

Darwin-18.7.0-x86_64-i386-64bit
Python 3.7.3 (default, Mar 27 2019, 16:54:48) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
surprise 1.0.6

**Comment by cuilongyin** [comment]
I figured it out ...

Replacing the similarity matrix with all zeros is basically predicting with global means always. replacing with ones: taking the first k neighbors encountered, and predicting with their weighted means. Replacing with random numbers: predicting with random neighbors' weighted average. 

If the dataset is small, and the ratings are not statistically variable enough, the RMSE can be as low as 0.96 even with random similarity matrix. This is different from just recommending randomly since the predictions are always related to either the global mean or certain neighbors, in which case the RMSE should be much lower than 1.4

Anyway the code itself has no problem...
